wsmaster/che-core-api-devfile/README.md

```diff
@@ -0,0 +1,187 @@
+Devfile can only contain one tool with `cheEditor` type.
+
+```
+...
+tools:
+  - name: theia-editor
+    type: cheEditor
+    id: org.eclipse.che.editor.theia:1.0.0
+```
+
+#### chePlugin
+Describes the plugin which used in workspace by defining it's id. 
+It is allowed to have several `chePlugin` tools.
+
+```
+...
+  tools:
+   - name: exec-plugin
+     type: chePlugin
+     id: che-machine-exec-plugin:0.0.1
+```
+
+Both types above using composite id, which is colon-separated id and version of plugin from Che Plugin registry.  
+List of available Che plugins and more information about registry can be found on https://github.com/eclipse/che-plugin-registry 
+
+
+
+#### kubernetes/openshift
+More complex tool type, which allows to apply configuration from kubernetes/openshift lists. Content of the tool may be provided either via `local` attribute which points to the file with tool content.
+Currently, devfile can only contain one tool with `kubernetes` or `openshift` type.
+```
+...
+  tools:
+    - name: mysql
+      type: kubernetes
+      local: petclinic.yaml
+      selector:
+        app.kubernetes.io/name: mysql
+        app.kubernetes.io/component: database
+        app.kubernetes.io/part-of: petclinic
+```
+Contents of the `local` file is currently read _ONLY_ if the devfile and local file both placed in the same GitHub repository. 
+So, alternatively, if you need to post devfile with such tools to REST API, contents of K8S/Openshift list can be put directly into devfile using `localContent` field:
+
+```
+...
+  tools:
+    - name: mysql
+      type: kubernetes
+      local: petclinic.yaml
+       localContent: |
+           kind: List
+           items:
+            -
+             apiVersion: v1
+             kind: Pod
+             metadata:
+              name: ws
+             spec:
+              containers:
+              ... etc
+```
+
+
+#### dockerimage
+Tool type which allows to define docker image based configuration of container in workspace.
+Devfile can only contain one tool with `dockerimage` type. 
+
+```
+ ...
+ tools:
+   - name: maven
+     type: dockerimage
+     image: eclipe/maven-jdk8:latest
+     volumes:
+       - name: maven-repo
+         containerPath: /root/.m2
+     env:
+       - name: ENV_VAR
+         value: value
+     endpoints:
+       - name: maven-server
+         port: 3101
+         attributes:
+           protocol: http
+           secure: 'true'
+           public: 'true'
+           discoverable: 'false'
+     memoryLimit: 1536M
+```
+
+### Commands expanded
+Devfile allows to specify commands set to be available for execution in workspace. Each command may contain subset of actions, which are related to specific tool, in whose container it will be executed.
+
+
+```
+ ...
+ commands:
+   - name: build
+     actions:
+       - type: exec
+         tool: mysql
+         command: mvn clean
+         workdir: /projects/spring-petclinic
+```
+
+### Live working examples
+
+    //TODO
```

Comment:
What the plan here?

Reply:
I've prepared couple of working examples (web-nodejs-sample, web-java-spring-petclinic) but che.openshift.io is not able to run any of them. (for me seems Che 7 workspaces not working here at all).
So they still currlently in my own repo, and after we will be able to run it we will move somewhere into public example repos and provide links here.

Reply:
@mshaposhnik go ahead even if they don't work on che.openshift.io. There is a known issue that prevent che 7 workspace to run on OSO starter cluster (cf. https://che.statuspage.io)

Reply:
Ok so there is 3 repo with exampes:
1) https://github.com/mshaposhnik/web-nodejs-sample  very simple  H/W example;
2) https://github.com/mshaposhnik/web-java-spring-petclinic  port to Che 7using devfile
3) https://github.com/mshaposhnik/hello-world-frontend-plugin  fork of one of Theia plugin with devfile.

Any ideas where to host them? Into `che-samples` maybe but indicate they're an Che 7 samples?

Reply:
What about `github.com/redhat-developers/devfile/samples`?

Reply:
Ok, lets merge this one, and after that i prepare examples & add them as separate PR.